### DISCUSSION TITLE
Interest for adding support for disk images with other sector sizes?

### DISCUSSION THREAD

**Discussion by ecm-pushbx** [dosemu2/dosemu2#1959] [Ideas]
Hi, as recently mentioned on the FreeDOS lists some versions of DOS have at least some support for sector sizes other than 512 Bytes. The FreeDOS kernel is not yet among those. My boot sector and initial loaders and debugger and instsect application are also written with support for different sector sizes, though this is untested as yet.

From my study of DOS data and file system structures I have determined that DOS can support a sector size between 32 B and 8 KiB inclusive.

The lower limit is to store one full (SFN) directory entry, albeit loaders (like the ROM-BIOS's or the code in an MBR) and file system drivers will have to load multiple consecutive sectors to access the boot sector or BPB in the case of smaller sectors. (My FAT32+FSIBOOT loader, and my debugger and instsect application, do support the multiple reads to support small sector sizes. The ldosmbr repo does not yet provide MBR loaders which do.)

The upper limit is because many recent versions of DOS stores the index of a directory entry within a sector as a byte, so 256 directory entries per sector is the maximum, resulting in 8 KiB rather than the somewhat common 4 KiB disks. For larger sector sizes, the kernel must allocate more than 512 B for transfer buffers of course.

Is there interest in merging a patch that would allow different sector sizes to be explicitly set for disk images? If so I may start to work on that.

**Comment by stsp** [comment]
Why is that needed, and what will the
patch do? Just change some helpers
(like mbr helper) to call int13 with smaller
blocks?

**Comment by andrewbird** [comment]
I guess it comes down to how likely Dosemu/FDPP is going to be presented with FAT filesystems that use non 512 byte sectors. From what the mkfs.fat manpage shows, within the FAT specification that could be 1024, 2048 and 4096. So where are these filesystems like to come from? I suppose a user could make an image file with larger sectors, but why would they? As far as I can see they'd only be raw disks and partitions where the user has a 4K drive like many of the newer SSDs. I think the case for FreeDOS support for it on real hardware is stronger.

**Reply by stsp** [reply]
Correct me if I am wrong, but I think
the larger sector size is just an optimization.
You don't have to match the physical
media's sector size, it may just be a bit
faster.
And as for smaller than 512b sectors -
I can't even imagine a reason.

**Reply by skitt** [reply]
Larger sector sizes were used in some OEM version of DOS to work around the 32M partition limit; see https://retrocomputing.stackexchange.com/q/26547/79 for details (in the question, the answers aren’t much use IMO).

**Reply by stsp** [reply]
I think this thread haven't settled on
a terminology just yet.
Yes, you can have FAT with alternative
sector size. But it seems ecm was talking
about MBR, int13 and, overall, an image
sector size.
So while I can understand the FAT sector
size, I still don't understand what exactly
does he mean here.

**Reply by ecm-pushbx** [reply]
>I think this thread haven't settled on
a terminology just yet.
Yes, you can have FAT with alternative
sector size.

Yes, and that's not what I meant. "Logically-sectored" typically seems to indicate that a layer other than int 13h translates between two different sector sizes.

> But it seems ecm was talking
about MBR, int13 and, overall, an image
sector size.
So while I can understand the FAT sector
size, I still don't understand what exactly
does he mean here.

That should read "she".

And yes, I'm referring to the int 13h (CHS and LBA) sector size.

**Reply by stsp** [reply]
> And yes, I'm referring to the int 13h (CHS and LBA) sector size.

Could you please point me to some
non-512b blocks LBA scheme? I've
never heard of such. I thought LBA
standard hard-codes 512b blocks.

**Comment by ecm-pushbx** [comment]
> Why is that needed, and what will the
patch do? Just change some helpers
(like mbr helper) to call int13 with smaller
blocks?

Smaller sector sizes will need that, yes, but the main change is just accessing the image data with the different size.

> I guess it comes down to how likely Dosemu/FDPP is going to be presented with FAT filesystems that use non 512 byte sectors. From what the mkfs.fat manpage shows, within the FAT specification that could be 1024, 2048 and 4096. So where are these filesystems like to come from? I suppose a user could make an image file with larger sectors, but why would they? As far as I can see they'd only be raw disks and partitions where the user has a 4K drive like many of the newer SSDs. I think the case for FreeDOS support for it on real hardware is stronger.

My main reason for this is adding and testing the support in my and FreeDOS applications/kernels.

**Reply by stsp** [reply]
> Smaller sector sizes will need that, yes, but the main change is just accessing the image data with the different size.

I still don't understand anything.
What int13 API depends on a sector size?
And what is "accessing the image data with the different size"?
Do you mean that the size of MBR/VBR
also changes, or what changes?

**Reply by andrewbird** [reply]
Int 13/48 reports the sector size in bytes http://www.ctyme.com/intr/rb-0715.htm. I wonder if that returned 4k, then int13/42 read one block would be 4k also http://www.ctyme.com/intr/rb-0708.htm.

**Reply by ecm-pushbx** [reply]
> I still don't understand anything.
What int13 API depends on a sector size?

All sector reads and writes, 13.02, 13.03, 13.42, and 13.43.

> And what is "accessing the image data with the different size"? Do you mean that the size of MBR/VBR
also changes, or what changes?

The MBR and VBR structures imply a certain minimum size which is why applications have to issue multiple consecutive accesses to work with partition tables, BPBs, etc on a smaller sector size. For larger sector sizes this is not a problem but then the buffers need to be large enough.

>Int 13/48 reports the sector size in bytes http://www.ctyme.com/intr/rb-0715.htm. I wonder if that returned 4k, then int13/42 read one block would be 4k also http://www.ctyme.com/intr/rb-0708.htm.

That is my expectation. However I'm not using 13.48 to get the sector size but rather I detect it using two reads of the very first sector, using differently initialised 8 KiB buffers, to determine how much is actually written by a 13.02 call (which assumes that the contents of that sector won't change between the two reads). This works even if the LBA extensions aren't supported by a specific disk.

**Reply by stsp** [reply]
> I wonder if that returned 4k, then int13/42 read one block would be 4k also http://www.ctyme.com/intr/rb-0708.htm.

It explicitly says "blocks", not sectors.
Have you ever seen a non-512 block
in some LBA standard?

**Reply by stsp** [reply]
So is the problem specific to those
old CHS int13 funcs?
I don't believe LBA is involved here,
you'd have to prove if it is.

**Reply by stsp** [reply]
> That is my expectation.

And my expectation is the opposite.
I've never seen such crap for LBA.
For CHS yes, but its legacy.

**Reply by andrewbird** [reply]
>  However I'm not using 13.48 to get the sector size but rather I detect it using two reads of the very first sector, using differently initialised 8 KiB buffers, to determine how much is actually written by a 13.02 call (which assumes that the contents of that sector won't change between the two reads).

This reply https://stackoverflow.com/questions/62483420/int-13h-effective-sector-size on stack overflow suggests that the int13 old read/write functions are fixed at 512 bytes, only the extended ones vary.

**Reply by stsp** [reply]
Because bios translates the real CHS
to the virtual representation to push the
size limits a bit. It increases cyls and
sectors, and sets the sector size to 512b
AFAIK.
But again, this wasn't always so.
Yet more legacy bioses worked with
real CHS, so this artifact about a sector
size is still there in a legacy funcs.

**Reply by ecm-pushbx** [reply]
Reading that page I don't see that they say CHS functions only ever use 512 B, just that they do in the absence of 13.48 call support. And someone else there mentions "4K native" disks. Anyway, you're saying it is LBA-only, stsp is saying it is CHS-only, but do you want me to add any of this as patches to dosemu2?

**Reply by stsp** [reply]
> Anyway, you're saying it is LBA-only, stsp is saying it is CHS-only,

Maybe @andrewbird (mis?)interpreted this part:
```
The extended BIOS disk services like [Int 13h/AH=42h](http://www.ctyme.com/intr/rb-0708.htm) and [Int 13h/AH=43h]
(http://www.ctyme.com/intr/rb-0710.htm) do not make any such assumption that a sector size is fixed at 512 bytes. On
any kind of drive that supports extended disk BIOS services you can query the drive parameters to determine the disk
sector size.
```

This is true, however those functions operate
with blocks, not sectors. They allow you to
_get_ the sector size, but they don't like using
it. And what kind of a sector size they reveal
I don't know, but that stackoverflow reply suggests
maybe the raw sector size it is.

> but do you want me to add any of this as patches to dosemu2?

Certainly not before we know how it works. :)

**Comment by ecm-pushbx** [comment]
> It explicitly says "blocks", not sectors.
Have you ever seen a non-512 block
in some LBA standard?

Block can mean many things, the list referring to "blocks" in a "Logical Block Addressing" function doesn't seem to me like it means something other than "sector".

> So is the problem specific to those
old CHS int13 funcs? I don't believe LBA is involved here,
you'd have to prove if it is.

If CHS uses a different sector size then LBA should use the same sector size, not some magic "always 512 B" size of what you call "blocks". In the list entry it explicitly says:

> 08h    QWORD   starting absolute block number
> (for non-LBA devices, compute as (Cylinder*NumHeads + SelectedHead) * SectorPerTrack + SelectedSector - 1)

Which wouldn't make sense if the LBA "block" and CHS "sector" could be of different sizes.

> And my expectation is the opposite. I've never seen such crap for LBA. For CHS yes, but its legacy.

Slightly different topic but we can also disable LBA as an option. Then this legacy crap could be implemented just for CHS accesses with other sector sizes and LBA disabled so as to force the applications to use CHS.

If you don't want to accept patches for this just say so. As mentioned I need this to develop and test support in the kernel, so I may or may not add it to a dosemu2 fork regardless, but I would certainly prefer to have it in everyone's dosemu2.

**Reply by andrewbird** [reply]
@ecm-pushbx I mentioned earlier that 4k blocks are commonly found on more modern SSDs, which is true but they mostly have something called '512e' that translates the physical 4k into logical 512b that older OSes can use. If you're looking for a test drive you can find that larger HDDs expose the sectors as '4kn' (native) which is why some machines have trouble booting from them.

**Reply by stsp** [reply]
> Which wouldn't make sense if the LBA "block" and CHS "sector" could be of different sizes.

You forgot that bios translates raw CHS
to virtual CHS. Unless that bios is very, very
old, it doesn't work with raw sector size.

> If you don't want to accept patches for this just say so.

I don't want to accept any patches I
don't understand. And in this particular
case I certainly don't understand what
and why do you want to change.

**Reply by stsp** [reply]
@andrewbird Agree.
So AFAICS there is nothing to change here.

**Reply by stsp** [reply]
> Slightly different topic but we can also disable LBA as an option. Then this legacy crap could be implemented
> just for CHS accesses with other sector sizes and LBA disabled so as to force the applications to use CHS.

This proposal is inconsistent.
If you can disable LBA then you assume
your bios supports LBA. In which case,
guaranteed, it also uses virtual CHS with
the 512b sector size.

**Reply by ecm-pushbx** [reply]
> You forgot that bios translates raw CHS to virtual CHS. Unless that bios is very, very old, it doesn't work with raw sector size.

I assume that the list comment I quoted is referring to the (13.08) "virtual" CHS. It doesn't matter to me what addressing mode the ROM-BIOS uses to talk to the hardware, I just think that it should use the 13.48 sector size regardless of if you use 13.02 or 13.42.

> I don't want to accept any patches I don't understand. And in this particular case I certainly don't understand what and why do you want to change.

I want the sector access functions to use sector sizes other than 512 B, and the backing image files to be addressed using these sizes, to emulate "4K native" formats or other unusual sector sizes, to test kernel support for those.

> I mentioned earlier that 4k blocks are commonly found on more modern SSDs, which is true but they mostly have something called '512e' that translates the physical 4k into logical 512b that older OSes can use. If you're looking for a test drive you can find that larger HDDs expose the sectors as '4kn' (native) which is why some machines have trouble booting from them.

It would be cheaper for me to just add this support to a dosemu2 branch, in both money and effort in connecting a new drive (and getting a system to connect it to).

**Reply by ecm-pushbx** [reply]
> If you can disable LBA then you assume
your bios supports LBA. In which case,
guaranteed, it also uses virtual CHS with
the 512b sector size.

I meant disable LBA as in instruct dosemu2 to not offer LBA on a specific disk. This is nothing to do with what an application does; if I disabled dosemu2 LBA then the application would *have* to use CHS or it would fail.

**Reply by ecm-pushbx** [reply]
To circle back to what you wrote @andrewbird

> If you're looking for a test drive you can find that larger HDDs expose the sectors as '4kn' (native) which is why some machines have trouble booting from them.

And why do they have trouble booting from them? Because the firmware or software doesn"t support other sector sizes. Which I interpret as the int 13h interfaces (CHS and LBA) would work with 4 KiB sectors/blocks in these cases.

**Reply by stsp** [reply]
> I just think that it should use the 13.48 sector size regardless of if you use 13.02 or 13.42.

Well, you can't start patching only
because you _think_ its the case, right?
We need a proof.
I unfortunately don't know what sector
size does 48 return. But I can only think
of the following possible scenarios:
- 48 returns raw sector size and the bios
  uses raw CHS and no LBA. On such
  bios what you say is true, but there are
  no such bioses around these days.
- 48 returns raw sector size and the bios
   is LBA-aware. All other funcs ignore that
   value and use 512b virtual sector size.
- 48 returns virtual sector size, which is
   always 512b.

If either of the above is correct, then no
patches are needed. If some other scenario
is real, it needs to be demonstrated.

**Reply by stsp** [reply]
> 48 returns raw sector size and the bios
> uses raw CHS and no LBA.

Except that w/o LBA support there is no 48.
And I don't know what other func can reveal
the sector size...
So it is not unlikely that bios always, even
in pre-LBA era, was translating the geometry
when using the non-512b media. And later
bioses inherited that technique for pushing
the size limit.
But in either case, after ATA-6, the sector
size was firmly defined to 512b. So even if
48 happens to report something else, you'd
better just ignore it.

**Reply by ecm-pushbx** [reply]
https://forum.osdev.org/viewtopic.php?f=1&t=29915

> You can use "[Int 0x13 extensions, Get Drive Parameters](http://www.ctyme.com/intr/rb-0715.htm)" to determine the bytes per sector for hard disks, CD-ROM, etc. This will be the sector size that "int 0x13" functions use (which can be completely different to what the underlying hardware uses); and because there's probably a lot of software that assumes 512-bytes (for hard disks) or 2048-bytes (for CD) it's relatively safe to assume BIOS emulates 512-bytes (for hard disks) or 2048-bytes (for CD) if/when the underlying hardware uses something different. For an OS (e.g. a native driver that can't use BIOS's emulation) you'd ask the device itself (e.g. the "Identify Device" ATA command).

...

> The only reliable way to determine what the BIOS does is to fill a large (16 KiB?) area with zeros and read a sector, then fill the area with 0xFFFFFFFF and read the same sector; so that you can determine how many of your bytes were overwritten.

https://forum.osdev.org/viewtopic.php?f=1&t=29915&start=15

>> On how much hardware can we assumer a sector to be of size 512 bytes for INT 13/AH=02h calls?
>
> Other than floppy disks, I think all devices you can read with that function will have 512B sectors.

**Reply by ecm-pushbx** [reply]
> And I don't know what other func can reveal
the sector size...

As quoted from the osdev forum and as I already described in https://github.com/dosemu2/dosemu2/discussions/1959#discussioncomment-5354900 you can check how much a 13.02 call actually writes, doing two calls with different fill patterns in the buffer to make sure you get the correct size.

> So it is not unlikely that bios always, even
in pre-LBA era, was translating the geometry
when using the non-512b media. And later
bioses inherited that technique for pushing
the size limit.

I don't think I am as confident in that assumption as you are.

**Reply by ecm-pushbx** [reply]
This comment does support your translation thesis, but it also says LBA functions may use a different sector size:

> Int 13h/AH=2h and AH=3h will always be emulated 512 no matter what the real sector size on the media. This however is not the case with the extended disk reads and writes via Int 13h/AH=42 and Int 13h/AH=43h. The sector size can vary. This also means that the amount of drive you can access may be more limited with the non-extended functions.

https://www.reddit.com/r/osdev/comments/ajfmtf/comment/ef323cb/?utm_source=share&utm_medium=mweb3x

**Reply by stsp** [reply]
> actually writes, doing two calls with different fill patterns in the buffer to make sure you get the correct size.

So you really suggest the one needs
to do all that to even just find out how
many bytes would the int13 call return?
Come on!

> This comment does support your translation thesis, but it also says LBA functions may use a different sector size:

I think these guys are all mis-guided
by the underspecced 0x48 fn. They
extrapolate it to other fns, that are
clearly documented to work with LBA
blocks, not CHS sectors.
Also I am pretty confident from the
first chapter of the ATA-6 spec (sorry,
the spec is huge, so I won't go past
the first chapter, but you can do that
to get the more consistent proofs) that
LBA-aware bioses work with sector
size of 512b, which likely (but not
certainly) means even 0x48 returns
512b always. Why "not certainly"?
Well, because of the silly EDD spec.
All other fields returned by 0x48 are
_underlined_ there as being the physical
value, but for sector size nothing is
said. Its quite hard to believe thought
that all fields are physical and sector
size is logical, so from EDD spec I
still deduce 0x48 can actually return
the raw value, but you should use it
only as an optimization, to align your
I/O to that margin.

**Comment by stsp** [comment]
LBA was invented in ATA-6 standard.
The quick search in a pdf reveals this:
```
3.1.44 sector: A uniquely addressable set of 256 words (512 bytes)
```
Page 5, chapter Definitions and abbreviations.

Which basically matches my picture of
the world.

**Comment by andrewbird** [comment]
I just asked this question over at fdisk issues list https://gitlab.com/FreeDOS/base/fdisk/-/issues/14, if the response is positive @ecm-pushbx perhaps there are possibilities to find out how dosemu's BIOS should behave if you want your disk to report as 4Kn.

**Reply by andrewbird** [reply]
It seems the reporter has a system that reports different geometry on int13/48 (apparently ATA values where the heads differ) and surmises that if he were to have a disk that was 4Kn then the sectors would be different too. I'm not sure this helps much.

**Comment by stsp** [comment]
What makes your wish to code that
up for FreeDOS if no one will ever
need or use that?
Have you got an indication from the
freedos ppl first that its needed at all?
Have you got any feature request from
the user that its needed at all?
I mean, what's your motivation?
Just plain WHY? :)

**Comment by ecm-pushbx** [comment]
I added a comment to https://superuser.com/questions/1271871/4k-emulation-sata-usb-controllers#comment2766975_1271871 :

> `lsusb` shows `Bus 009 Device 004: ID 174c:55aa ASMedia Technology Inc. ASM1051E SATA 6Gb/s bridge, ASM1053E SATA 6Gb/s bridge, ASM1153 SATA 3Gb/s bridge, ASM1153E SATA 6Gb/s bridge` for my SATA/USB adapter. For a 512 GB or 1 TB drive it uses 512 B/sector as usual. For a 4 TB drive it uses 4 KiB/sector (4Kn? or "4Ke"? GParted says 4 KiB) whereas the same drive when directly connected to the internal SATA controller runs in 512e mode. (`hdparm` does say physical sector size 4 KiB, logical 512 B then. GParted also says 512 B in this setup.) Same data either way, if read directly (eg `dd`).

I formatted a small FAT16 partition (15 MiB) on my Linux box's system 120 GB SSD, sorted in the MBR partition table as `/dev/sda4` (or `hda4` in lDebug notation). I copied `lcdebug.com` to this partition, and installed an lDOS boot loader to the boot sector using `instsect.com`. Then I made the Linux go into suspend-hybrid sleep. (Don't worry, after the experiments I was able to load up the hibernated Linux system again.) After switching the power off and on again, `grub` boots. I enter "C" to get to a command line. I enter the commands:

```
set root=(hd0,4)
insmod chain
chainloader +1
boot
```

This successfully boots into lCDebug. Next I tried out interrupt 13h services 02h and 48h. Result: The 4 KiB/sector 4 TB drive as provided by my USB/SATA adapter does not show up anywhere. Two USB devices do show up, those correspond to the two slots of my USB-connected card reader. They return size in sectors and Bytes/sector both as 0, and trying to read results in an error 31h (no medium).

Now for the most interesting part, connecting a 1 TB drive to the same adapter, and booting into lCDebug again, does make it so the drive shows up and is accessible on int 13h. In this case the adapter works with 512 B/sector.

The 4 KiB drive does not show up in the firmware's setup or boot selection screen, either. The 512 B drive does show up there.

So that adventure hints somewhat at what you claim, that int 13h should always use 512 B/sector. At least my Linux box's firmware doesn't allow me to access a 4 KiB/sector drive with its CSM USB support.

**Reply by ecm-pushbx** [reply]
The "ASRock UEFI Setup Utility" identifies as "UEFI Version: FM2A88X Extreme4+ P2.90". And when I checked that out Linux did fail to wake up from suspend-hybrid. Meh!

**Comment by stsp** [comment]
That's good to know.
I would still expect to see some bioses
that return something else via 0x48 fn
as EDD spec emphasizes on the "physical"
values for all fields except "sector size",
but it can't be virtual while others are
physical. So at least that should be possible,
but no extrapolation on any other function
should be done.
